server/test/integration/dbTest/beginTransactionTest.js

```diff
@@ -0,0 +1,112 @@
+  })
+
+  describe('when beginTransaction fails the first time because of a deadlock, then successfully connects after retry', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables to reflect the retry', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction is successful the first time it is called', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sinon.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.logError).to.not.be.called
+    })
+
+    it('should not log a retry', () => {
+      expect(helpers.log).to.not.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction fails twice, deadlocking on the first attempt and throwing an error with a valid pgClient on the second attempt', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      clientStub.query.withArgs('BEGIN').rejects(new Error('some error'))
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying beginTransaction.`)
+    })
+
+    it('should log the second error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the beginTransaction query: Error: some error`)
+    })
+
+    it('should return null', () => {
+      expect(pgClient).to.be.null
+    })
+  })
+})
```

Comment:
Great tests!

Reply:
Minor question; should these tests be under `integration` or `unit`? My understanding is that unit tests stub surrounding logic so as to observe only the functionality of some unit (class, function), whereas integration tests check the functionality of some unit as it integrates with another unit. Correct my understanding if I'm wrong, but I feel like these tests are unit tests and not integration tests?

Reply:
Very fair question! I actually asked Theresa the same thing once I split the functionality into two functions, and she said it would be considered integration tests, because we are hitting the exported `beginTransaction`(since that's what we want to work) as opposed to just testing `runBeginTransactionWithRetries`.

Reply:
OK, cool!